src/bootstrap/config.rs

```diff
@@ -782,14 +771,75 @@ impl Config {
         }
 
+        // This is duplicated in bootstrap.py.
+        config.out = build.build_dir.unwrap_or_else(|| PathBuf::from("build"));
+        t!(fs::create_dir_all(&config.out));
+        config.out = config.out.canonicalize().expect("Build directory canonicalized");
+        config.out = config
+            .out
+            .components()
+            .flat_map(|component| {
+                if component.as_os_str() == "$ROOT" {
+                    config.src.components().collect::<Vec<_>>()
+                } else {
+                    vec![component]
+                }
+            })
+            .collect();
+
+        if config.dry_run {
+            let dir = config.out.join("tmp-dry-run");
+            t!(fs::create_dir_all(&dir));
+            config.out = dir;
+        }
+
+        let build_rustc = build.rustc;
+        config.initial_rustc = env::var_os("RUSTC")
+            .map(PathBuf::from)
+            .or_else(|| {
+                // Try a rustc binary by the CARGO. Cargo doesn't provide an
+                // absolute path to us in $RUSTC for build scripts.
+                let rustc = config.initial_cargo.with_file_name(exe("rustc", config.build));
+                if rustc.exists() {
+                    if let Some(file_cfg) = build_rustc.clone().map(PathBuf::from) {
+                        assert_eq!(rustc, file_cfg.canonicalize().expect("rustc exists"));
+                    }
+                    Some(rustc)
+                } else {
+                    build_rustc.clone().map(PathBuf::from)
+                }
+            })
+            .or_else(|| build_rustc.clone().map(PathBuf::from))
+            .unwrap_or_else(|| panic!("Unable to discover a rustc binary"));
+
+        // Cargo does not provide a RUSTFMT environment variable, so we
+        // synthesize it manually. Note that we also later check the config.toml
+        // and set this to that path if necessary.
+        let rustfmt = config.initial_rustc.with_file_name(exe("rustfmt", config.build));
+        config.initial_rustfmt = if rustfmt.exists() { Some(rustfmt) } else { None };
+
         // Now that we've reached the end of our configuration, infer the
         // default values for all options that we haven't otherwise stored yet.
 
-        set(&mut config.initial_rustc, build.rustc.map(PathBuf::from));
-        set(&mut config.initial_cargo, build.cargo.map(PathBuf::from));
+        // Make sure that our variables, pulled from Cargo, are equivalent to
+        // configuration.
+        if let Some(cargo) = build.cargo.map(PathBuf::from) {
+            assert_eq!(config.initial_cargo, cargo.canonicalize().expect("cargo exists"));
+        }
 
         config.llvm_skip_rebuild = llvm_skip_rebuild.unwrap_or(false);
 
+        // OUT_DIR is a path like $target/debug/build/bootstrap-xxx. We want
+        // target/debug.
+        config.rustbuild_bin_directory = PathBuf::from(env!("OUT_DIR"))
```

Comment:
Instead of using `OUT_DIR` here (which makes assumptions about the output directory structure Cargo may wish to change one day), could this use `env::current_exe()` and go one level up from that?